miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/storage/MiniAppSecureStorage.kt

```diff
@@ -0,0 +1,257 @@
+                } ?: kotlin.run {
+                    onFailed(MiniAppStorageError.emptyStorage)
+                }
+            } else {
+                onFailed(MiniAppStorageError.unavailableStorage)
+            }
+            storageState.postValue(StorageState.UNLOCK)
+        }
+    }
+
+    fun load(
+        miniAppId: String,
+        onSuccess: (Map<String, String>) -> Unit,
+        onFailed: (MiniAppStorageError) -> Unit
+    ) {
+        scope.launch {
+            storageState.postValue(StorageState.LOCK)
+            if (isStorageAvailable(miniAppId)) {
+                val storedItems = readFromEncryptedFile(miniAppId)
+                storedItems?.let {
+                    onSuccess(storedItems)
+                } ?: kotlin.run {
+                    onFailed(MiniAppStorageError.emptyStorage)
+                }
+            } else {
+                onFailed(MiniAppStorageError.unavailableStorage)
+            }
+            storageState.postValue(StorageState.UNLOCK)
+        }
+    }
+
+    fun insertItems(
+        miniAppId: String,
+        items: Map<String, String>,
+        onSuccess: (Map<String, String>) -> Unit,
+        onFailed: (MiniAppStorageError) -> Unit
+    ) {
+        scope.launch {
+            storageState.postValue(StorageState.LOCK)
+            writeToEncryptedFile(miniAppId, serializedItems(items), onSuccess, onFailed)
+            storageState.postValue(StorageState.UNLOCK)
+        }
+    }
+
+    fun getItem(
+        miniAppId: String,
+        key: String,
+        onSuccess: (String) -> Unit,
+        onFailed: (MiniAppStorageError) -> Unit
+    ) {
+        scope.launch {
+            if (isStorageAvailable(miniAppId)) {
+                val storedItems = readFromEncryptedFile(miniAppId)
+                storedItems?.let {
+                    if (storedItems.containsKey(key)) {
+                        onSuccess(storedItems[key] ?: "")
+                    } else {
+                        onFailed(MiniAppStorageError.unavailableItem)
+                    }
+                } ?: kotlin.run {
+                    onFailed(MiniAppStorageError.unavailableItem)
+                }
+            } else {
+                onFailed(MiniAppStorageError.unavailableStorage)
+            }
+        }
+    }
+
+    private fun deserializeItems(jsonToRead: String): Map<String, String> {
+        return Gson().fromJson(
+            jsonToRead,
+            object : TypeToken<Map<String, String>>() {}.type
+        )
+    }
+
+    private fun serializedItems(items: Map<String, String>): String = Gson().toJson(items)
+
+    /**
+     * Will be invoked by MiniApp.clearSecureStorage(miniAppId: String).
+     * @param miniAppId will be used to find the file to be deleted.
+     */
+    fun clearSecureStorage(miniAppId: String) {
+        if (isStorageAvailable(miniAppId)) {
+            scope.launch {
+                val file = File(secureStorageBasePath, "$miniAppId.txt")
+                file.delete()
```

Comment:
It's possible that the Host App could call `clearSecureStorage` while the Mini App is still open. What would happen in this case?

Reply:
@corycaywood we always create a new encrypted file when `miniapp` try to set new items, because encrypted files data can not be overwritten. if host app clear the storage while the `miniapp` is open, it will just simply create a new file and store the data.